plugins/catalog-backend/src/ingestion/processors/AwsS3DiscoveryProcessor.ts

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2020 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { UrlReader } from '@backstage/backend-common';
+import { LocationSpec } from '@backstage/catalog-model';
+import limiterFactory from 'p-limit';
+import * as result from './results';
+import {
+  CatalogProcessor,
+  CatalogProcessorEmit,
+  CatalogProcessorParser,
+} from './types';
+
+export class AwsS3DiscoveryProcessor implements CatalogProcessor {
+  constructor(private readonly reader: UrlReader) {}
+
+  async readLocation(
+    location: LocationSpec,
+    optional: boolean,
+    emit: CatalogProcessorEmit,
+    parser: CatalogProcessorParser,
+  ): Promise<boolean> {
+    if (location.type !== 's3-discovery') {
+      return false;
+    }
+
+    try {
+      const output = await this.doRead(location.target);
+      for (const item of output) {
+        for await (const parseResult of parser({
+          data: item.data,
+          location: { type: location.type, target: item.url },
+        })) {
+          emit(parseResult);
+        }
+      }
+    } catch (error) {
+      const message = `Unable to read ${location.type}, ${error}`;
+
+      if (error.name === 'NotFoundError') {
+        if (!optional) {
+          emit(result.notFoundError(location, message));
+        }
+      } else {
+        emit(result.generalError(location, message));
+      }
+    }
+    return true;
+  }
+
+  private async doRead(
+    location: string,
+  ): Promise<{ data: Buffer; url: string }[]> {
+    const limiter = limiterFactory(5);
+    const response = await this.reader.readTree(location);
```

Comment:
So I think my only remaining question here is, can this be used at all in buckets with mixed contents? The processor's use case now, is limited to having a bucket with cataog-info files and nothing else, is that right? For a future extension, maybe it would make sense to be able to specify `target: https://testbucket.s3.us-east-2.amazonaws.com/foo/bar/**/catalog-info.yaml` for example, and we only ever try to parse things that actually match that glob.

Reply:
Sweet! Thank you and the greater backstage community for the opportunity to contribute!

Yes, as far as I am aware, it can be used in buckets with mixed contents. We do not make a check in `readTree()` for any specific file extensions. For now our use case will have a bucket containing only catalog-info files.

Reply:
I think the problem happens later; the part of the code that then tries to parse every file that came out of readTree will find it problematic to parse a jpeg. :)

Reply:
Ahh, good point. Would it be useful then to have a check in the processor to ensure what is parsed ends in `.yaml` or other appropriate file extensions?

Reply:
I'd suggest the ability to ask for `https://x.com/root/folder/**/catalog-info.yaml` for example where asterisk glob matching is supported. It would then `readTree` from `https://x.com/root/folder` and filter the files out of that by matching against the glob pattern.

If no pattern is given at all, it could be assumed to be `**/catalog-info.yaml`, for example. Or your org might choose just `*.yaml` if they are all named differently in a flat list.

There are glob libraries to use, I think we use `minimatch` or `globby` in some places already? Have to check